base/services/w32time/w32time.c

```diff
@@ -0,0 +1,418 @@
+                                          0);
+                }
+            }
+        }
+        CloseHandle(hToken);
+    }
+
+    return Ret;
+}
+
+
+/*
+ * NTP servers state the number of seconds passed since
+ * 1st Jan, 1900. The time returned from the server
+ * needs adding to that date to get the current Gregorian time
+ */
+static DWORD
+UpdateSystemTime(ULONG ulTime)
+{
+    FILETIME ftNew;
+    LARGE_INTEGER li;
+    SYSTEMTIME stNew;
+
+    /* Time at 1st Jan 1900 */
+    stNew.wYear = 1900;
+    stNew.wMonth = 1;
+    stNew.wDay = 1;
+    stNew.wHour = 0;
+    stNew.wMinute = 0;
+    stNew.wSecond = 0;
+    stNew.wMilliseconds = 0;
+
+    /* Convert to a file time */
+    if (!SystemTimeToFileTime(&stNew, &ftNew))
+    {
+        return GetLastError();
+    }
+
+    /* Add on the time passed since 1st Jan 1900 */
+    li = *(LARGE_INTEGER *)&ftNew;
+    li.QuadPart += (LONGLONG)10000000 * ulTime;
+    ftNew = * (FILETIME *)&li;
+
+    /* Convert back to a system time */
+    if (!FileTimeToSystemTime(&ftNew, &stNew))
+    {
+        return GetLastError();
+    }
+
+    if (!SystemSetTime(&stNew))
+    {
+        return GetLastError();
+    }
+
+    return ERROR_SUCCESS;
+}
+
+
+static DWORD
+GetIntervalSetting(VOID)
+{
+    HKEY hKey;
+    DWORD dwData;
+    DWORD dwSize = sizeof(dwData);
+    LONG lRet;
+
+    dwData = 0;
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                      L"SYSTEM\\CurrentControlSet\\Services\\W32Time\\TimeProviders\\NtpClient",
+                      0,
+                      KEY_QUERY_VALUE,
+                      &hKey) == ERROR_SUCCESS)
+    {
+        /* This key holds the update interval in seconds
+         * It is useful for testing to set it to a value of 10 (Decimal)
+         * This will cause the clock to try and update every 10 seconds
+         * So you can change the time and expect it to be set back correctly in 10-20 seconds
+         */
+        lRet = RegQueryValueExW(hKey,
+                                L"SpecialPollInterval",
+                                NULL,
+                                NULL,
+                                (LPBYTE)&dwData,
+                                &dwSize);
+        RegCloseKey(hKey);
+    }
+
+    if (lRet != ERROR_SUCCESS)
+        return 0;
+    else
+        return dwData;
+}
+
+
+int
+SetTime(VOID)
+{
+    ULONG ulTime;
+    LONG lRet;
+    HKEY hKey;
+    DWORD dwIndex = 0;
+    DWORD dwValSize;
+    WCHAR szValName[MAX_VALUE_NAME];
+    WCHAR szData[MAX_VALUE_NAME];
+    DWORD cbName = sizeof(szData);
+    WCHAR szDefault[MAX_VALUE_NAME] = L"";
+    size_t Size;
+
+    DPRINT("Entered SetTime.\n");
+
+    lRet = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                         L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\DateTime\\Servers",
+                         0,
+                         KEY_QUERY_VALUE,
+                         &hKey);
+    if (lRet != ERROR_SUCCESS)
+    {
+        return 1;
+    }
+
+    while (TRUE)
```

Comment:
It looks like this loop could simply be a RegQueryValueEx call?

Reply:
I believe that you are correct. I modeled this after some other code that must have needed to do the enumerate to search through the keys, but here we know we need the "default" one.